systemd: enable homed

kloenk opened the pull request:
###### Motivation for this change
Fixes #91243
Contains #98998 (merged)

###### Things done
Please also see https://github.com/NixOS/nixpkgs/pull/72802#issuecomment-695337602

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

CC: @Mic92 (changed your systemd patch)

FRidh commented on pkgs/os-specific/linux/systemd/default.nix:
```diff
@@ -1,24 +1,61 @@
 let
   version = "246";
+
+  trueFalse =  cond: if cond then "true" else "false";
```
`lib.boolToString`

flokli commented:
Can you add a minimal test for the new functionality?

We would never want this to silently regress...

Kloenk commented:
@flokli you mean for systemd-homed? first I have to understad it :-). I'm currently trying to rebuild my pc, and then figure out how homed works.

flokli commented on pkgs/top-level/all-packages.nix:
```diff
@@ -12134,7 +12134,22 @@ in
   dbxml = callPackage ../development/libraries/dbxml { };
 
-  dbus = callPackage ../development/libraries/dbus { };
+  dbus = callPackage ../development/libraries/dbus { systemd = systemd.override {
+      withResolved = false;
+      withLogind = false;
+      withHostnamed = false;
+      withLocaled = false;
+      withTimedated = false;
+      withHwdb = false;
+      withEfi = false;
+      withImportd = false;
+      withHomed = false;
+      withPortabled = false;
+      dbus = null;
+      cryptsetup = null;
+      lvm2 = null;
+    };
```
Huuuh, so we pull in another systemd here, to bootstrap dbus, which is being used by systemd?

Did you check if it'll become a runtime dependency?

I'm not sure if this is the right path forward - also see https://github.com/NixOS/nixpkgs/issues/98094 for another bootstrap systemd leaking in :-/

Kloenk replied:
It's a runtime dependency, yes. Maybe we could patchelf dbus to search in /run/current-system for libsystemd

arianvp replied:
Systemd's DBUS dependency is _only_ used in its test suite (which we disable); so we should be able to just set `dbus = null` on our toplevel `systemd` instead of providing a `systemdMinimal` 

Source:

https://github.com/systemd/systemd/blob/master/meson_options.txt#L338-L339

flokli commented on pkgs/top-level/all-packages.nix:
```diff
@@ -18120,7 +18135,21 @@ in
     # which depends on lvm2 again.  But we only need the libudev part
     # which does not depend on cryptsetup.
-    udev = udev.override { cryptsetup = null; };
+    udev = udev.override {
```
Is this a separate, or the same that's passed into `dbus` above? We should give it an attribute name and refer it in both places.

Kloenk replied:
It's the same. Something like systemdMinimal?

flokli replied:
:+1:

Kloenk replied:
To what should udev point? systemd or systemdMinimal?

Kloenk commented:
Should I add `systemd-homed.service` to `upstreamSystemUnits` in `nixos/modules/system/boot/systemd.nix`?
Not exactly sure how I would enable that unit if I include it in `upstreamSystemUnits`.

lheckemann commented:
Suggestion: add a `Co-Authored-By` mention of @kirelagin to acknowledge his initial work on this :)

flokli commented:
@Kloenk is this ready for review?

We might want to degrade this to a draft PR until it has proven to work, and has a test ensuring it stays that way ;-)

Kloenk commented:
Oh, yeah. It's a draft.
But I think I can't change it right now, as I only have my phone with me.

Kloenk commented:
I think the way to continue with this should be:
- [ ] figuring out how and what can be used from systemd-homed
- [ ] write a nixos module, which then enables the systemd-homed service
- [ ] write tests

arianvp commented on pkgs/top-level/all-packages.nix:
```diff
@@ -12212,7 +12212,7 @@ in
   dbxml = callPackage ../development/libraries/dbxml { };
 
-  dbus = callPackage ../development/libraries/dbus { };
+  dbus = callPackage ../development/libraries/dbus { systemd = systemdMinimal; };
```
As noted on an outdated part of the PR (sorry for that),

Instead of depending `systemdMinimal` lets just remove the `dbus` dependency on our toplevel `systemd` package. As it's only used in the test suite (which afaik we disable anyway).    Then we do not need two copies of systemd in our closure.  

(Though we still have to keep it for `udev` I guess; but that's a different issue, not related to this PR)

Kloenk replied:
I'm pretty sure systemd-homed needs dbus

arianvp replied:
That would really surprise me. But I'm happy to be proven wrong.   All systemd services usually use the built-in `sd-bus` library for dbus support, for bootstrapping reasons.

arianvp replied:
Also searching through the source code; only a test seems to import `dbus/dbus.h` :

https://github.com/systemd/systemd/search?q=%22dbus%2Fdbus.h%22&type

So homed requiring `libdbus` seems very unlikely to me..

Kloenk replied:
You seem to be right. I had some problems compiling homed in the beginning, and they did go away as I added dbus and some other libraries. Will see what happens without dbus (I think it should work).

Kloenk replied:
removed dbus as input from systemd. do we still want to keep systemdMinimal?

arianvp replied:
I'd say no. As it's not really related to your PR anymore. It cleans up the diff a bit.

flokli replied:
I'd prefer to keep it - due to https://github.com/NixOS/nixpkgs/pull/98299#discussion_r495240873 and as a alternative to https://github.com/NixOS/nixpkgs/pull/97051.

Kloenk replied:
I also used it to compile libfido2

Kloenk commented:
> I think the way to continue with this should be:
> 
>     * [ ]  figuring out how and what can be used from systemd-homed
> 
>     * [ ]  write a nixos module, which then enables the systemd-homed service
> 
>     * [ ]  write tests

I installed that systemd on my main machine xD.
I can now create users via `homectl create test` and see via `userdbctl user test` that it was created via homectl. Is this a good thing to test?

arianvp commented on pkgs/os-specific/linux/systemd/default.nix:
```diff
@@ -1,22 +1,57 @@
+, kmod
+
+# Optional dependencies
+, pam, cryptsetup, lvm2, audit, acl
```
We could/should probably add optional dependencies for `libp11-kit` and `libfido2` .  Both can be used to unlock systemd-homed partitions using your yubikey. which is one of the cool main features of homed.

The first one provides support to unlock using PKCS11, whilst the other uses the more modern FIDO2 protocol for crypto.


However this might potentially add an interesting cycle; as `libfido2` depends on `udev`.  So we might need to build `udev` with `  { libfido2 = null; }`

flokli replied:
And wouldn't that be a reason to keep `systemdMinimal`?

Kloenk replied:
is `libp11-kit` in `libp11`?

arianvp replied:
Nope it's in `nixpkgs.p11-kit`

Kloenk replied:
systemd-homework now links against libfido2 and libp11-kit. So I think it works

Kloenk commented:
systemd-homed needs this tools in scope, should we always include them, or make a nixos module to enable them?
`mkfs.xfs` mkfs.btrfs` and `mkfs.ext4`

flokli commented:
> systemd-homed needs this tools in scope, should we always include them, or make a nixos module to enable them?
> `mkfs.xfs` mkfs.btrfs`and`mkfs.ext4`

What does "in scope" mean here? `$PATH`? Does it go via `systemd-makefs.service`, or is a different logic?

There's similar problems for `x-systemd.makefs`, `x-systemd.growfs`, and I doubt hardcoding paths to the specific tools will do us a favour. If these calls are dispatched via `systemd-makefs.service`, and `systemd-makefs.service` can be persuaded to look in `$PATH`, we should have a way to add this to its PATH while assembling NixOS, without needing it for building.

Kloenk commented:
Yes, as far as I have seen it we just should be able to add the tools to the path of systemd-homed.service. my question is, should everything be included, or should there be a nixos option to enable and disable those filesystems?

flokli commented on nixos/modules/security/pam.nix:
```diff
@@ -318,6 +318,14 @@ let
+        default = config.systemd.homed.enable;
+        type = types.bool;
+        description = ''
+          If enabled, pam_systemd_homed will attempt to unlock te homed user.
```
```suggestion
          If enabled, pam_systemd_homed will attempt to unlock homed users.
```

Kloenk replied:
added you for this as Co-Author

flokli commented on nixos/modules/system/boot/systemd-homed.nix:
```diff
@@ -0,0 +1,55 @@
+    systemd.homed = {
+      enable = mkEnableOption "Enable homed";
+
+      DefaultStorage = mkOption {
```
```suggestion
      defaultStorage = mkOption {
```

IMHO this would be more consistent with other nixpkgs options.

Kloenk replied:
Made this big, because then it's the same spelling as in the file. Like systemd units also use big letters.

flokli replied:
Well, NixOS module options usually start with small letters. The things starting with big letters are usually members of attrsets (and not real NixOS options) rendered by Nix.

flokli commented on nixos/modules/system/boot/systemd-homed.nix:
```diff
@@ -0,0 +1,55 @@
+        '';
+      };
+
+      DefaultFileSystemType = mkOption {
```
same

flokli commented:
> Yes, as far as I have seen it we just should be able to add the tools to the path of systemd-homed.service. my question is, should everything be included, or should there be a nixos option to enable and disable those filesystems?

A good starting point might be `config.system.fsPackages`, even though it's heavily influenced by `boot.supportedFilesystems`, which might not be enough.

We can't really know what filesystem a homed user wants to create/mount before they try.

flokli commented:
@Kloenk any chance you could you the `#nixos-systemd` IRC channel?

Kloenk commented:
My irc setup is kinda broken at the moment. Will try to get my weechat to join there later today

arianvp commented on pkgs/top-level/all-packages.nix:
```diff
@@ -12944,7 +12944,7 @@ in
   herqq = libsForQt514.callPackage ../development/libraries/herqq { };
 
-  hidapi = callPackage ../development/libraries/hidapi { };
+  hidapi = callPackage ../development/libraries/hidapi { systemd = systemdMinimal; };
```
I just checked the source code, and `hidapi` and `libusb1` actually depends on `udev`; not `systemd` (though of course those are one and the same package).   For documentation sake it would be nice to name the parameters `udev` instead of `systemd`; and then you don't need to repeat `systemdMinimal` all over the place but just need the toplevel `udev = systemdMinimal` alias.

If that's something you're quickly willing to do; I think it would be a lot clearer for the next reader. But also feel free to ignore this as bikeshedding.  It's not a blocker by any means

flokli replied:
Pointing `udev` to `systemdMinimal` would be a logical consequence, but this might cause quite some rebuilds and possibly break other stuff, as it hasn't been used properly/consistently throughout nixpkgs. This should definitely be another PR.

Ericson2314 replied:
Another PR is fine --- i doubt these changes are necessary for this one at all, but I think we can still rename the variable before the big change, giving us:
```suggestion
  hidapi = callPackage ../development/libraries/hidapi { udev = systemdMinimal; };
```
and separately worry about making `udev = systemdMinimal` everywhere.

Ericson2314 replied:
My suggestion is now done for both those packages.

flokli commented on nixos/modules/security/pam.nix:
```diff
@@ -419,6 +431,8 @@ let
           # Password management.
           password sufficient pam_unix.so nullok sha512
+          ${optionalString cfg.enableHomed
+          "-password [success=1 default=ignore] ${pkgs.systemd}/lib/security/pam_systemd_home.so"}
```
I'm not sure about the right parameters and order here - this still seems to be somewaht WIP:

https://github.com/systemd/systemd/issues/16906

Kloenk replied:
I took this from an arch vm. Not exactly sure how it works, but it seems to work under arch

flokli commented on pkgs/os-specific/linux/systemd/default.nix:
```diff
@@ -85,9 +124,10 @@ in stdenv.mkDerivation {
+      [ libffi audit lz4 bzip2 libapparmor iptables ] ++
+      stdenv.lib.optional withEfi gnu-efi ++
+      stdenv.lib.optional withSelinux libselinux ++
+      stdenv.lib.optionals withHomed [ cryptsetup.dev p11-kit libfido2 ];
```
We don't want cryptsetup only when homed is enabled - This is also generally useful to have systemd understand (and act upon) `/etc/crypttab` on - see https://github.com/NixOS/nixpkgs/pull/66856

Maybe adding another `withCryptsetup` flag that sets `-Dcryptsetup=true`, and is a requirement for `withHomed`?

Looking at https://github.com/NixOS/nixpkgs/pull/97008/files#diff-036410e9211b4336186fc613f7200b12R18035, it seems to be fine if this would be set to `false` for `systemdMinimal`.

chandradeepdey commented on nixos/modules/security/pam.nix:
```diff
@@ -337,6 +345,8 @@ let
           # Account management.
           account required pam_unix.so
+          ${optionalString cfg.enableHomed
+            "-account [success=1 default=ignore] ${pkgs.systemd}/lib/security/pam_systemd_home.so"}
```
This makes `pam_acct_mgmt()` equivalent to ```int pam_acct_mgmt() {
    return PAM_SUCCESS;
}```
We have the same problem in Arch. I would suggest either patching systemd or following https://github.com/systemd/systemd/issues/16906#issuecomment-692088553 for minimum (one) loopholes.

Kloenk replied:
Can you give me an example of how to configure this? I'm not that good with pam

chandradeepdey replied:
> I would suggest either patching systemd or following [systemd/systemd#16906 (comment)](https://github.com/systemd/systemd/issues/16906#issuecomment-692088553) for minimum (one) loopholes.

patching systemd would be just removing [L836](https://github.com/systemd/systemd/blob/092f5d2ed34e0a72995cd6808ba7bf1c535561ec/src/home/pam_systemd_home.c#L836) and [L837](https://github.com/systemd/systemd/blob/092f5d2ed34e0a72995cd6808ba7bf1c535561ec/src/home/pam_systemd_home.c#L837). Then you can follow https://github.com/systemd/systemd/issues/16906#issuecomment-683439776 (this is a different comment from the one quoted above).

flokli replied:
I'd suggest awaiting the outcome of https://github.com/systemd/systemd/issues/16906 - this should really be fixed upstream.

flokli replied:
Also, our own PAM infrastructure is somewhat of a mess currently: https://github.com/NixOS/nixpkgs/issues/90640

flokli replied:
Apparently, this should have been fixed in https://github.com/systemd/systemd/pull/17376, which should have landed in systemd 247.

flokli replied:
The PAM refactor is ongoing in https://github.com/NixOS/nixpkgs/pull/105319.

kloenk committed:
systemd: Add an option for homed

This is disabled by default to indicate that is hasn't been adiquately
tested with NixOS yet.

kloenk committed:
top-level: Use `systemdMinimal` in a few more places

Kloenk commented:
I'm currently in the process of rebuilding it on top of #98998

Ericson2314 commented on pkgs/os-specific/linux/systemd/default.nix:
```diff
@@ -49,9 +51,13 @@ assert withImportd ->
   && gnutar != null && gnupg != null);
 
+# Homed needs cryptsetup
+assert withHomed -> ( withCryptsetup != false );
```
```suggestion
assert withHomed -> withCryptsetup;
```
`x : x != false` is the identity function for booleans :).

doronbehar commented:
So, this change only adds homed support for the `systemd` package but doesn't introduce new nixos options right? I'd like to experiment with homed support for my Laptop at some point, but I will not introduce a `rebuild-stdenv` change to my nixpkgs fork. Likely, even if I our systemd had homed support, only time would have allowed me to learn this new feature well enough so that I maybe could have written a good nixos module. What if we'd merge this PR as is, but without officially announcing homed as supported, at least so that users could experiment with it? After some experience will be gained in the community, we'd be able to write good tests to validate it works between channel updates.

arianvp commented:
You can set `systemd.package` in your NixOS config to avoid rebuilding the world.

It just means that packages will still link to the systemd without homed; but I think this is fine enough for testing.

arianvp commented:
> Should I add systemd-homed.service to upstreamSystemUnits in nixos/modules/system/boot/systemd.nix?
Not exactly sure how I would enable that unit if I include it in upstreamSystemUnits.

We can leave it our for now; and let people explicitly enable it with `systemd.additionalUpstreamUnits`

flokli commented:
I'd *really* prefer if we'd first address the PAM issues mentioned by Archlinux, and include a test, before shipping this in master.

arianvp commented:
PAM configuration are a NixOS Module concern no?  This PR does not do
anything related to NixOS modules so far; and I think the point of Doron
was that if the systemd package has homed support; then it's easier for
others to contribute a working NixOS module (including the  correct PAM
configuration).

On Mon, Nov 2, 2020 at 7:03 PM Florian Klink <notifications@github.com>
wrote:

> I'd *really* prefer if we'd first address the PAM issues mentioned by
> Archlinux, and include a test, before shipping this in master.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/98299#issuecomment-720633790>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEZNI3MFA7L7MLRPNP4AC3SN3X5LANCNFSM4RTIF2WA>
> .
>

kisik21 commented:
I agree on including this as a package. The module can be contributed later, and those who feel brave enough could try the new tech on their machines or VMs and help contributing the aforementioned module.

I would totally install this on a VM for testing!

flokli commented:
Can we add a `enableHomed` flag, which still is set to `false` in our default `systemd` derivation?

That'd mean we won't ship untested features enabled by default, but people can get a systemd with homed, without rebuilding the world, by slightly overriding `systemd.package`.

Ericson2314 commented:
Pity draft PRs don't get their conflict status updated. This certainly has cosmetic conflicts with #101886.

Ericson2314 commented:
Also, do we have an issue on our end for the PAM debacle?

Kloenk commented:
Will try to rebase and recheck tomorrow

Kloenk commented:
Rebased, not yet tested if it builds and works

flokli commented:
I successfully build both `systemd` and `systemd.overrideAttrs ( old: { withHomed = true; } )`.

flokli reviewed (approved):
This LGTM. This should unblock some more testing and toying with it without too many world rebuilds :-) Happy to flip the default once there's a module system integration and a VM test.

Ericson2314 reviewed (approved):
OK applied my nit-pick and then cleanued up history to separate the bootstrapping changes from the changes to systemd proper. Looks all good to me now!

flokli commented:
⛵🚀

doronbehar commented on pkgs/top-level/all-packages.nix:
```diff
@@ -13126,7 +13126,10 @@ in
 
-  hidapi = callPackage ../development/libraries/hidapi { };
+  hidapi = callPackage ../development/libraries/hidapi {
+    # TODO: remote once `udev` is `systemdMinimal` everywhere.
```
*remove

Ericson2314 replied:
```suggestion
    # TODO: remove once `udev` is `systemdMinimal` everywhere.
```

doronbehar commented on pkgs/top-level/all-packages.nix:
```diff
@@ -14394,6 +14399,8 @@ in
   libusb1 = callPackage ../development/libraries/libusb1 {
     inherit (darwin) libobjc;
     inherit (darwin.apple_sdk.frameworks) IOKit;
+    # TODO: remote once `udev` is `systemdMinimal` everywhere.
```
Same here - *remove

Ericson2314 replied:
```suggestion
    # TODO: remove once `udev` is `systemdMinimal` everywhere.
```

doronbehar reviewed (commented):
Looks good to me besides that comment typo.
